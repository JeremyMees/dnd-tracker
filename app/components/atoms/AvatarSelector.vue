<script setup lang="ts">
const emit = defineEmits<{ update: [string] }>()

const props = defineProps<{
  identifier: string
  options: string[]
  selected?: string | number
}>()

const current = computed<number>(() => {
  return props.selected
    ? props.options.findIndex((option: string) => option === props.selected)
    : 0
})

const icon = computed<string>(() => {
  switch (props.identifier) {
    case 'backgroundColor': return 'tabler:background'
    case 'accessories': return 'tabler:eyeglass-2'
    case 'clothingColor': return 'tabler:shirt'
    case 'face': return 'tabler:lego'
    case 'facialHair': return 'tabler:moustache'
    case 'head': return 'tabler:mood-boy'
    default: return 'tabler:palette'
  }
})

function prev(): void {
  const index = current.value === 0 ? props.options.length - 1 : current.value - 1
  const option = props.options[index]

  if (option) emit('update', option)
}

function next(): void {
  const index = current.value === props.options.length - 1 ? 0 : current.value + 1
  const option = props.options[index]

  if (option) emit('update', option)
}
</script>

<template>
  <div class="text-foreground flex items-center justify-between gap-2 px-2 border-4 rounded-lg bg-secondary border-secondary min-w-[160px] overflow-hidden">
    <Icon
      v-tippy="$t(`general.${identifier === 'head' ? 'hair' : identifier}`)"
      data-test-icon
      :name="icon"
      class="min-w-6 min-h-6"
      aria-hidden="true"
    />
    <span
      data-test-current
      class="w-8 text-xs text-muted-foreground"
    >
      {{ current + 1 }}/{{ options.length }}
    </span>
    <div class="flex items-center">
      <button
        data-test-prev
        :aria-label="$t('actions.prev')"
        class="w-5 flex items-center justify-center focus:outline-primary"
        @click="prev"
      >
        <Icon
          class="min-w-6 min-h-6"
          name="tabler:caret-left"
          aria-hidden="true"
        />
      </button>
      <button
        data-test-next
        :aria-label="$t('actions.next')"
        class="w-5 flex items-center justify-center focus:outline-primary"
        @click="next"
      >
        <Icon
          class="min-w-6 min-h-6"
          name="tabler:caret-right"
          aria-hidden="true"
        />
      </button>
    </div>
  </div>
</template>
