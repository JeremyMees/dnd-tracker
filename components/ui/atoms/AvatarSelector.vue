<script setup lang="ts">
const emit = defineEmits<{ update: [string] }>()

const props = defineProps<{
  identifier: string
  options: string[]
  selected?: string | number
}>()

const { t } = useI18n()

const current = computed<number>(() => {
  return props.selected
    ? props.options.findIndex((option: string) => option === props.selected)
    : 0
})

const icon = computed<string>(() => {
  switch (props.identifier) {
    case 'backgroundColor': return 'tabler:background'
    case 'accessories': return 'tabler:eyeglass-2'
    case 'clothingColor': return 'tabler:shirt'
    case 'face': return 'tabler:lego'
    case 'facialHair': return 'tabler:moustache'
    case 'head': return 'tabler:mood-boy'
    default: return 'tabler:palette'
  }
})

function prev(): void {
  const index = current.value === 0
    ? props.options.length - 1
    : current.value - 1

  emit('update', props.options[index])
}

function next(): void {
  const index = current.value === props.options.length - 1
    ? 0
    : current.value + 1

  emit('update', props.options[index])
}
</script>

<template>
  <div class="flex items-center justify-between gap-2 px-2 btn-small-black min-w-[160px] overflow-hidden">
    <button
      :aria-label="t('actions.prev')"
      class="min-w-7 min-h-7 flex items-center justify-center focus:outline-primary"
      @click="prev"
    >
      <Icon
        class="min-w-6 min-h-6"
        name="tabler:chevron-left"
        aria-hidden="true"
      />
    </button>
    <Icon
      v-tippy="t(`general.${identifier === 'head' ? 'hair' : identifier}`)"
      :name="icon"
      class="min-w-6 min-h-6"
      aria-hidden="true"
    />
    <span class="w-10">
      {{ current + 1 }}/{{ options.length }}
    </span>
    <button
      :aria-label="t('actions.next')"
      class="min-w-7 min-h-7 flex items-center justify-center focus:outline-primary"
      @click="next"
    >
      <Icon
        class="min-w-6 min-h-6"
        name="tabler:chevron-right"
        aria-hidden="true"
      />
    </button>
  </div>
</template>
