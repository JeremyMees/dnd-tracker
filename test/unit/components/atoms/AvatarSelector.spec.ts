import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'
import AvatarSelector from '~/components/atoms/AvatarSelector.vue'

type Identifier = 'backgroundColor' | 'accessories' | 'clothingColor' | 'face' | 'facialHair' | 'head' | 'color'

interface Props {
  identifier: Identifier
  options: string[]
  selected?: string | number
}

const identifiers: Record<Identifier, { label: string, icon: string }> = {
  backgroundColor: { label: 'backgroundColor', icon: 'tabler:background' },
  accessories: { label: 'accessories', icon: 'tabler:eyeglass-2' },
  clothingColor: { label: 'clothingColor', icon: 'tabler:shirt' },
  face: { label: 'face', icon: 'tabler:lego' },
  facialHair: { label: 'facialHair', icon: 'tabler:moustache' },
  head: { label: 'head', icon: 'tabler:mood-boy' },
  color: { label: 'color', icon: 'tabler:palette' },
}

const props: Props = {
  identifier: 'backgroundColor',
  options: ['1', '2'],
}

describe('AvatarSelector', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(AvatarSelector, { props })
    expect(component.html()).toMatchSnapshot()
  })

  it('Should render avatar with default props correctly', async () => {
    const component = await mountSuspended(AvatarSelector, { props })

    const icon = component.find('[data-test-icon]')
    expect(icon.exists()).toBeTruthy()
    expect(icon.attributes('class')).toContain(identifiers[props.identifier].icon)

    const current = component.find('[data-test-current]')
    expect(current.exists()).toBeTruthy()
    expect(current.text()).toBe(`1/${props.options.length}`)
  })

  it('Should go to next and previous options', async () => {
    const component = await mountSuspended(AvatarSelector, { props })

    const options = props.options.length
    const nextButton = component.find('[data-test-next]')
    const prevButton = component.find('[data-test-prev]')

    expect(component.find(`[data-test-current]`).text()).toBe(`1/${options}`)

    // First click on next, should go to option 2
    await nextButton.trigger('click')
    let emitted = component.emitted('update')
    expect(emitted).toBeTruthy()
    component.setProps({ selected: emitted?.[0]?.[0] as string })
    await nextTick()
    expect(component.find(`[data-test-current]`).text()).toBe(`2/${options}`)

    // Second click on next, should wrap around to option 1
    await nextButton.trigger('click')
    emitted = component.emitted('update')
    expect(emitted).toBeTruthy()
    component.setProps({ selected: emitted?.[1]?.[0] as string })
    await nextTick()
    expect(component.find(`[data-test-current]`).text()).toBe(`1/${options}`)

    // Click on prev, should go to option 2
    await prevButton.trigger('click')
    emitted = component.emitted('update')
    expect(emitted).toBeTruthy()
    component.setProps({ selected: emitted?.[2]?.[0] as string })
    await nextTick()
    expect(component.find(`[data-test-current]`).text()).toBe(`2/${options}`)
  })
})
