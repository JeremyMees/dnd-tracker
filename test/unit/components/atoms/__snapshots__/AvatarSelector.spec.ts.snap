// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AvatarSelector > Should match snapshot 1`] = `
"<div class="text-foreground flex items-center justify-between gap-2 px-2 border-4 rounded-lg bg-secondary border-secondary min-w-[160px] overflow-hidden"><span class="iconify i-tabler:background min-w-6 min-h-6" aria-hidden="true" data-test-icon=""></span><span data-test-current="" class="w-8 text-xs text-muted-foreground">1/2</span>
  <div class="flex items-center"><button data-test-prev="" aria-label="actions.prev" class="w-5 flex items-center justify-center focus:outline-primary"><span class="iconify i-tabler:caret-left min-w-6 min-h-6" aria-hidden="true"></span></button><button data-test-next="" aria-label="actions.next" class="w-5 flex items-center justify-center focus:outline-primary"><span class="iconify i-tabler:caret-right min-w-6 min-h-6" aria-hidden="true"></span></button></div>
</div>"
`;
